lib/icingadb/icingadb.cpp

```diff
@@ -67,21 +67,29 @@ void IcingaDB::Start(bool runtimeCreated)
 	m_Rcon = new RedisConnection(GetHost(), GetPort(), GetPath(), GetPassword(), GetDbIndex(),
 		GetEnableTls(), GetInsecureNoverify(), GetCertPath(), GetKeyPath(), GetCaPath(), GetCrlPath(),
 		GetTlsProtocolmin(), GetCipherList(), GetDebugInfo());
-	m_Rcon->SetConnectedCallback([this](boost::asio::yield_context& yc) {
+
+	auto connectedCallback ([this](boost::asio::yield_context& yc) {
 		m_WorkQueue.Enqueue([this]() { OnConnectedHandler(); });
 	});
+
+	m_Rcon->SetConnectedCallback([this, connectedCallback](boost::asio::yield_context& yc) {
+		for (auto& kv : m_Rcons) {
+			kv.second->Start();
+		}
+
+		m_Rcon->SetConnectedCallback(connectedCallback);
```

Comment:
Why is this needed? Only the first call to `OnConnectedHandler()` should do anything anyways (there's a check in there and on any subsequent call, it should just return).

Reply:
It’s named `OnConnectedHandler`. A programmer who works on all this expects it being called on every connect. This PR's goal isn’t to make the code as complicated as possible, but to only change _that one thing_.

Reply:
I don't understand the claim that this PR is supposed to change only one thing. Before, `OnConnectHandler` was only called for the parent connection, so isn't this changing another thing?

By the way, the name `OnConnectedHandler` was chosen when there was only a single connection. So I think it would also be fine to just rename it to better reflect the current purpose instead of complicating the code for the sake of its name. But isn't this really another construction area?

Reply:
Before, all Redis connections were started at the same time. After, they’re started by the m_ConnectedCallback, but only the first m_ConnectedCallback() time. So yes, only one thing.

Reply:
Upon closer inspection: Once `OnConnectedHandler()` is called, Icinga starts with the initial dump using *all* connections, including the child connections that are started at the very same time. Is there anything that ensures that queries issued to the child connections won't get lost because they're not yet ready? Or would the correct behavior actually be to only call `OnConnectedHandler()` after all connections are ready?

Reply:
They won’t get lost. They’ll get queued.

Reply:
Well, yes, there's a queue. But looks like there's nothing that prevents the execution of these jobs before the connection is ready. This also explains the mysterious errors I asked about in our internal chat a few days ago.

Using this patch:
```diff
diff --git a/lib/icingadb/redisconnection.hpp b/lib/icingadb/redisconnection.hpp
index fff9a0036..aed339afa 100644
--- a/lib/icingadb/redisconnection.hpp
+++ b/lib/icingadb/redisconnection.hpp
@@ -13,6 +13,7 @@
 #include "base/string.hpp"
 #include "base/tlsstream.hpp"
 #include "base/value.hpp"
+#include "base/logger.hpp"
 #include <boost/asio/buffer.hpp>
 #include <boost/asio/buffered_stream.hpp>
 #include <boost/asio/deadline_timer.hpp>
@@ -257,7 +258,7 @@ private:
 class RedisDisconnected : public std::runtime_error
 {
 public:
-       inline RedisDisconnected() : runtime_error("")
+       inline RedisDisconnected() : runtime_error("redis disconnected")
        {
        }
 };
@@ -270,7 +271,7 @@ public:
 class RedisProtocolError : public std::runtime_error
 {
 protected:
-       inline RedisProtocolError() : runtime_error("")
+       inline RedisProtocolError() : runtime_error("redis protocol error")
        {
        }
 };
@@ -368,6 +369,7 @@ void RedisConnection::WriteOne(StreamPtr& stream, RedisConnection::Query& query,
        namespace asio = boost::asio;
 
        if (!stream) {
+               Log(LogWarning, "IcingaDB") << "Lost Query: " << query[0];
                throw RedisDisconnected();
        }
 
```

I receive this:
```
[2021-07-27 17:01:42 +0200] information/IcingaDB: Starting initial config/status dump
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:notificationcommand:customvar' 'icinga:notificationcommand:envvar' 'icinga:checksum:notificationcommand:envvar' 'icinga:notificationcommand:argument' 'icinga:checksum:notificationcommand:argument' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:eventcommand:customvar' 'icinga:eventcommand:envvar' 'icinga:checksum:eventcommand:envvar' 'icinga:eventcommand:argument' 'icinga:checksum:eventcommand:argument' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:servicegroup:customvar' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:zone:customvar' 'icinga:zone:parent' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:user:customvar' 'icinga:usergroup:member' 'icinga:user:state' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: MULTI
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'MULTI' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:usergroup:customvar' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:timeperiod:customvar' 'icinga:timeperiod:override:include' 'icinga:timeperiod:override:exclude' 'icinga:timeperiod:range' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: MULTI
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'MULTI' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:host:customvar' 'icinga:hostgroup:member' 'icinga:host:state' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: DEL
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Error during sending query 'DEL' 'icinga:endpoint:customvar' which has been fired and forgotten: redis disconnected
[2021-07-27 17:01:42 +0200] warning/IcingaDB: Lost Query: HSCAN
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] critical/IcingaDB: Exception during ConfigDump: redis disconnected
[2021-07-27 17:01:42 +0200] information/IcingaDB: Initial config/status dump finished in 0.205472 seconds.
```

Reply:
Other construction area. To be precise: #8939